### PR TITLE
Add the support of GreyBox in MindtPy

### LINKED ISSUE #2767: Use cyipopt with Mindtpy
## Summary

When using the GreyBox model in Pyomo, we get an error message saying cyipopt cannot be used with mindtpy. 

### Description

Error message: 
Traceback (most recent call last):
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 1262, in _cast
    return self._domain(value)
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 249, in __call__
    raise ValueError("value %s not in domain %s" % (value, self._domain))
ValueError: value cyipopt not in domain ['ipopt', 'gams', 'baron']

During handling of the above exception, another exception occurred:

```
Traceback (most recent call last):
  File "rotary-bed-cost-strategy.py", line 439, in <module>
    results = solver.solve(mod, mip_solver='gurobi_persistent', nlp_solver='cyipopt', tee=True, integer_tolerance=0.5)
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/contrib/mindtpy/MindtPy.py", line 79, in solve
    config.set_value(kwds)
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 2193, in set_value
    self[key] = value[_decl_map[key]]
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 2010, in __setitem__
    self._data[_key].set_value(val)
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 1652, in set_value
    self._data = self._cast(value)
  File "/home/parallels/anaconda3/envs/pyomo/lib/python3.7/site-packages/pyomo/common/config.py", line 1273, in _cast
    (self.name(True), value, _dom, err))
ValueError: invalid value for configuration 'nlp_solver':
	Failed casting cyipopt
	to <class 'pyomo.common.config.In'>
	Error: value cyipopt not in domain ['ipopt', 'gams', 'baron']
```

### PR DESCRIPTION
**Pull request by ZedongPeng**
## Summary/Motivation:
This Pull Request adds the support of the Greybox in MindtPy.

## Changes proposed in this PR:
- Generates OA cuts using the information from the GreyBox model.
- Fixes a bug in [pyomo_grey_box_nlp.py](https://github.com/Pyomo/pyomo/blob/main/pyomo/contrib/pynumero/interfaces/pyomo_grey_box_nlp.py)
### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/algorithm_base_class.py:1337]: The use of this transformation is new. Is it motivated by our new tests? Is the transformation well-supported? @emma58

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/cut_generation.py:217]: Should this become an issue?

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (bernalde) [inline pyomo/contrib/mindtpy/cut_generation.py:230]: These cuts assume that the black box is: 1. in the objective, 2. is convex. For the 1st one, can we generalize it? For the second one, can we use an option to decide whether or not we want to generate the augmented penalty cuts? Our examples/tests satisfy both assumptions, but we can get some flexibility for not so much effort.

File: pyomo/contrib/pynumero/interfaces/pyomo_grey_box_nlp.py
Comment (bernalde) [inline pyomo/contrib/pynumero/interfaces/pyomo_grey_box_nlp.py:431]: Leave a comment on why are we adding `active=True` here and at least let's run this through @michaelbynum to see if it is a good change

**Review by bernalde — CHANGES_REQUESTED** [review]
Good work, Zedong! I request some changes and also some revisions from our friends that motivated this development! @adowling2 !

**Comment by bernalde** [comment]
The tests are failing because:
`FAILED pyomo/contrib/mindtpy/tests/test_mindtpy.py::TestMindtPy::test_OA_APPSI_ipopt - RuntimeError: A feasible solution was not found, so no solution can be loaded.Please set opt.config.load_solution=False and check results.termination_condition and results.best_feasible_objective before loading a solution.`
This has to do with the new option passed. Let's solve this first

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: `Gurobi` will fail even when we deactivate the `greybox` block. Therefore, we cannot use `Gurobi` as the mip solver.

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:1337]: After MindtPy rewrite, we repeatedly use the fixed-nlp subproblem. Therefore, we will call `'contrib.deactivate_trivial_constraints'` transformation and revert it after solving. I missed the feasibility subproblem and this might be a bug.

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:230]: In this implementation, we don't assume the grey box is in the objective function. It will also work if it is in the constraints. 
The constraints inside the grey box are nonlinear equality constraints. If the constraint is convex, we should use **Equality Relaxation**. If the constraint is nonconvex, we should use both **Equality Relaxation** and **Augmented Penalty**. We can add slack variables using `config.add_slack`, the same as algebraic constraints. 
Therefore, our current implementation works even without the two assumptions.

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: I will submit an issue related to this.

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: https://github.com/Pyomo/pyomo/issues

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: @emma58 Could you provide me a `reference` example?

Comment (emma58) [reply pyomo/contrib/mindtpy/cut_generation.py:217]:
```
from pyomo.environ import *

m = ConcreteModel()
m.b = Block()
m.b.x = Var(bounds=(2, 5))
m.y = Var(domain=NonNegativeReals)
m.c = Constraint(expr=m.b.x - m.y <= 4)
m.obj = Objective(expr=m.b.x + m.y)

m.b.deactivate()

m.somewhere_safe_to_keep_things = Block()
m.somewhere_safe_to_keep_things.x = Reference(m.b.x)

opt = SolverFactory('gurobi_persistent')
opt.set_instance(m)

opt.solve(m, tee=True)
```
This is the main idea. You'll want to figure out how to do your own accounting in terms of finding the Vars you need to reference and naming the references, etc.

Comment (bernalde) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: #3000 Issue opened

Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: @emma58 . I don't think `Reference` is an elegant way to resolve this issue. I will comment the code of adding `Gurobi` lazy constraints and only support `greybox` in multi-tree implementation this time. 
Btw, since persistent solver will be totally replaced by `appsi_solver`, does appsi_solver support both `CPLEX` and `Gurobi` callbacks as `persistent solver` now? @michaelbynum

**Comment by ZedongPeng** [comment]
This PR is ready for review. @emma58

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (michaelbynum) [reply pyomo/contrib/mindtpy/cut_generation.py:217]: appsi_gurobi does support callbacks. appsi_cplex will, but does not currently.

**Comment by bernalde** [comment]
I don't think we're ready yet, Zedong.
Some errors from the tests:
```
=========================== short test summary info ============================
FAILED pyomo/contrib/mindtpy/tests/test_mindtpy.py::TestMindtPy::test_OA_APPSI_ipopt - RuntimeError: A feasible solution was not found, so no solution can be loaded.Please set opt.config.load_solution=False and check results.termination_condition and results.best_feasible_objective before loading a solution.
FAILED pyomo/environ/tests/test_environ.py::TestPyomoEnviron::test_tpl_import_time - AssertionError: Items in the first set but not the second:
'scipy' : Unexpected module found in 5 slowest-loading TPL modules
==== 2 failed, 12120 passed, 5009 skipped, 29 xfailed in 954.63s (0:15:54) =====
```

We are responsible for the first, and scipy taking so much longer seems to be a new thing. Let's aim to fix the first error and then we can keep trying.

**Comment by ZedongPeng** [comment]
Hi @jsiirola . `scipy` cannot be imported from Pyomo dependency. How can I fix this?

Import code
`from pyomo.common.dependencies.scipy.sparse import coo_matrix`

Output
```
  File "/home/runner/work/pyomo/pyomo/pyomo/contrib/pynumero/interfaces/external_grey_box.py", line 15, in <module>
    from pyomo.common.dependencies.scipy.sparse import coo_matrix
  File "/home/runner/work/pyomo/pyomo/pyomo/common/dependencies.py", line 168, in __getattr__
    return getattr(_mod, attr)
  File "/home/runner/work/pyomo/pyomo/pyomo/common/dependencies.py", line 75, in __getattr__
    raise DeferredImportError(self._moduleunavailable_message())
pyomo.common.errors.DeferredImportError: The scipy.sparse module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'scipy'
```

**Comment by jsiirola** [comment]
@ZedongPeng, you want to do something like:
```python
import pyomo.common.dependencies.scipy.sparse as scipy_sparse

# then later...
return scipy_sparse.coo_matrix((data, (row, col)), shape=(1, 5))
```

**Comment by ZedongPeng** [comment]
The error is not inside `MindtPy` now. How can we deal with it? @jsiirola 
```
=========================== short test summary info ============================
FAILED pyomo/environ/tests/test_environ.py::TestPyomoEnviron::test_tpl_import_time - AssertionError: Items in the first set but not the second:
'scipy' : Unexpected module found in 5 slowest-loading TPL modules
==== 1 failed, 12121 passed, 5009 skipped, 29 xfailed in 1001.61s (0:16:41) ====
```

**Comment by jsiirola** [comment]
Ahh - the issue is that mindtpy is importing PyNumero (in `pyomo/contrib/mindtpy/algorithm_base_class.py`), which is importing scipy.

You can see this by running `python pyomo/environ/tests/test_environ.py`.  You will see (in a lot of other output):
```
TPLS:
                            __future__:    179  (from pyomo.core.expr.visitor)
[....]
                          scipy.sparse:  12384  (from pyomo.contrib.pynumero.interfaces.external_grey_box)
                           scipy.stats: 134511  (from pyomo.contrib.pynumero.dependencies)
```

You can resolve things by deferring the import of pynumero using the `attempt_import` mechanism.  Replace the pynumero import with:
```python
from pyomo.common.dependencies import attempt_import

egb = attempt_import('pyomo.contrib.pynumero.interfaces.external_grey_box')[0]
```
and then replace instances of `ExternalGreyBoxBlock` with `egb.ExternalGreyBoxBlock`.

**Review by bernalde — CHANGES_REQUESTED** [review]
There seem to be some test failing still.
`E   pyomo.common.errors.DeferredImportError: The pyomo.contrib.pynumero.interfaces.external_grey_box module (an optional Pyomo dependency) failed to import: ModuleNotFoundError: No module named 'scipy'
------- generated xml file: /home/runner/work/pyomo/pyomo/TEST-pyomo.xml -------
=========================== short test summary info ============================
ERROR pyomo/contrib/mindtpy/tests/test_mindtpy.py
ERROR pyomo/contrib/mindtpy/tests/test_mindtpy_ECP.py
ERROR pyomo/contrib/mindtpy/tests/test_mindtpy_feas_pump.py
ERROR pyomo/contrib/mindtpy/tests/test_mindtpy_grey_box.py
ERROR pyomo/contrib/mindtpy/tests/test_mindtpy_lp_nlp.py
!!!!!!!!!!!!!!!!!!! Interrupted: 5 errors during collection !!!!!!!!!!!!!!!!!!!!
================== 30 skipped, 5 errors in 100.53s (0:01:40) ===================`
These tests fail on the slim verrsion and pypy versions (those without numpy). We are importing numpy instead of attempting that import

**Comment by ZedongPeng** [comment]
We should find a chance to fix all the `import` in `pyomo`.

**Comment by emma58** [comment]
@ZedongPeng, checking the slim tests and the pypy-3.9 ones, it looks like you need to skip some tests when numpy and scipy aren't available.

**Comment by ZedongPeng** [comment]
Do we have to put `import` at the front of the file? @emma58 
If so, we cannot skip the `scipy import error`. And the `scipy import error` is from `pynumero` instead of `MindtPy`.
```
error: cannot format /home/runner/work/pyomo/pyomo/pyomo/contrib/mindtpy/tests/test_mindtpy_grey_box.py: Cannot parse: 20:0: from pyomo.contrib.mindtpy.tests.MINLP_simple import SimpleMINLP as SimpleMINLP
```

**Comment by emma58** [comment]
Sorry, I mislead you, It looks like your test skips were okay before. My guess is that the test failures had to do with someplace that you import something from pynumero that is in turn importing numpy. Remember that when you use the deferred imports, the package gets imported when you use it, so you will probably still need to check if numpy is available and, if not, throw an error from mindtpy, before you import what you need from pynumero. I'll mark a couple places in the review where I think this might be happening.

File: pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py:9]: Objective function.

File: pyomo/contrib/mindtpy/tests/test_mindtpy_grey_box.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/tests/test_mindtpy_grey_box.py]: Got it.

File: pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py]: I changed the default value of verbose to `False`.

**Comment by bernalde** [comment]
Test failures point to import of scipy when it is not available. A conditional attempt import should resolve it.

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:216]: If we don't call `enumerate` over `values`, any suggestions about how to implement it? Directly `enumerate` over `variable`?

Comment (emma58) [reply pyomo/contrib/mindtpy/cut_generation.py:216]: I don't really know--you would know better than me. It's completely possible that this is fine. My concern is that you will end up with the wrong coefficient (`jacobian_matrix[index][var_index]`) paired with the wrong `var - value(var)` term if the `target_model_greyboox.inputs` dictionary is every constructed in a different order that doesn't correspond to the indices in `jacobian_matrix`. I don't know greybox well at all, so maybe it makes a promise that what I just said will never happen, in which case your code is totally fine. I'm just asking if you know if that's the case or not. It's always worth thinking carefully when you are relying on the order things come up when iterating over a dictionary...

File: pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py
Comment (emma58) [reply pyomo/contrib/mindtpy/tests/MINLP_simple_grey_box.py:9]: Ah. Could you write it out in the docstring?

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:2247]: Yes. This option should not be up to the users. Is there a way to define an option (not open to users) instead of config (open to users) in Pyomo?

Comment (emma58) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:2259]: It's not really an option at all, just something you calculate based on some of the config options, right? You could probably store it privately on the algorithm class. Just make sure to restore state afterwards if you do keep it on the class.

**Comment by ZedongPeng** [comment]
Now all the tests passed except the Jenkins one. Could you show me the output of Jenkins? @emma58 Thanks.

**Comment by emma58** [comment]
@ZedongPeng, I'm not convinced those failures are actually your fault... I just merged main, and we'll see what happens. It looks like you still have a few comments above to address though?

File: pyomo/contrib/mindtpy/algorithm_base_class.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:2259]: Is it a good choice to change `load_solutions` to the attribute of the Solver Object?

Comment (emma58) [reply pyomo/contrib/mindtpy/algorithm_base_class.py:2259]: That sounds fine to me. You should probably make it private as well. The only risk of doing this is that you should be careful to restore the state after the solve.

File: pyomo/contrib/mindtpy/cut_generation.py
Comment (ZedongPeng) [reply pyomo/contrib/mindtpy/cut_generation.py:216]: Currently, I cannot come up with a good way to improve this since the `evaluate_jacobian_outputs` returns a `scipy.sparse._coo.coo_matrix`.

Comment (emma58) [reply pyomo/contrib/mindtpy/cut_generation.py:216]: Maybe just put in a comment that you are relying on the order in which items are added to the `target_model_greybox.inputs` dictionary, so that future you (or someone else) will have a hint if this ever doesn't work.

**Comment by ZedongPeng** [comment]
I think this PR is ready to merge if all the tests pass.

**Review by emma58 — APPROVED** [review]

**Comment by emma58** [comment]
This is ready for a second review. @jsiirola, @mrmundt, @blnicho, @michaelbynum, any volunteers?

**Review by blnicho — APPROVED** [review]

**Pull request merged**
